### PR TITLE
v14: Refactor and enhance `System.Text.Json` converters

### PR DESCRIPTION
**Pull request by ronaldbarendse**
This refactors and enhances the `System.Text.Json` converters provided by the CMS, most of which were migrated from `Newtonsoft.Json` in PRs https://github.com/umbraco/Umbraco-CMS/pull/15728 and https://github.com/umbraco/Umbraco-CMS/pull/15685. The next version of Umbraco Deploy will also use STJ and the changes in this PR enhances these built-in converters, reducing the amount of duplicate code.

To summarize, I've done the following:
- Update `JsonUdiConverter` to support both `Udi` and `GuidUdi` (removing `JsonGuidUdiConverter`), plus now also adding support for `StringUdi` (used in Deploy for languages, templates and other file based entities);
- Rename `JsonBoolConverter` to `JsonFuzzyBooleanConverter` and update the implementation to always require a boolean (like) value (removing the fallback to `false` for invalid values);
- Add `ReadOnlyJsonConverter<T>` and `WriteOnlyJsonConverter<T>` base converters;
- Rename and improve the following converters:
  - `AutoInterningStringConverter` to `JsonStringInternConverter`;
  - `AutoInterningStringKeyCaseInsensitiveDictionaryConverter<TValue>` to `JsonDictionaryStringInternIgnoreCaseConverter<TValue>`;
  - `CaseInsensitiveDictionaryConverter<TValue>` to `JsonDictionaryStringIgnoreCaseConverter<TValue>`;
  - `ForceUtcDateTimeConverter` to `JsonUniversalDateTimeConverter`;
  - Aligned the MessagePack formatters by renaming `MessagePackAutoInterningStringKeyCaseInsensitiveDictionaryFormatter<TValue>` to `MessagePackDictionaryStringInternIgnoreCaseFormatter<TValue>`;
- Rename `SystemTextJsonSerializer` to `DefaultJsonSerializer` and `SystemTextConfigurationEditorJsonSerializer` to `DefaultConfigurationEditorJsonSerializer`;
- Add `JsonUdiRangeConverter` to support `UdiRange`;
- And finally also added the `sealed` keyword to all converters.

### CODE REVIEW

File: src/Umbraco.Infrastructure/Serialization/JsonStringInternConverter.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+/// Converts a string to or from JSON, interning the string when reading.
+/// </summary>
+public sealed class JsonStringInternConverter : ReadOnlyJsonConverter<string>
+{
+    /// <inheritdoc />
+    public override string? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        => reader.GetString() is string value
+        ? string.Intern(value)
+        : null;
+}
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Serialization/JsonStringInternConverter.cs:15]: Maybe this entire converter should be removed? string.Intern has many downsides: https://github.com/dotnet/runtime/issues/28368#issuecomment-453204870

File: src/Umbraco.Infrastructure/Serialization/JsonFuzzyBooleanConverter.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Umbraco.Cms.Infrastructure.Serialization;
+
+/// <summary>
+/// Converts a boolean (like) value to or from JSON.
+/// </summary>
+public sealed class JsonFuzzyBooleanConverter : JsonConverter<bool>
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Serialization/JsonFuzzyBooleanConverter.cs:9]: I don't think this name makes any sense. Fuzzy matching means matching something approximately, there's nothing approximately about this.

File: src/Umbraco.Infrastructure/Serialization/JsonObjectConverter.cs [hunk 3]
```diff
@@ -1,43 +1,42 @@
...
+        if (value is null)
         {
             return;
         }
 
         // If an object is equals "new object()", Json.Serialize would recurse forever and cause a stack overflow
         // We have no good way of checking if its an empty object
         // which is why we try to check if the object has any properties, and thus will be empty.
-        if (objectToWrite.GetType().Name is "Object" && !objectToWrite.GetType().GetProperties().Any())
+        if (value.GetType().Name is "Object" && !value.GetType().GetProperties().Any())
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Serialization/JsonObjectConverter.cs:28]:
```suggestion
        if (value.GetType() == typeof(object) && !value.GetType().GetProperties().Any())
```

File: src/Umbraco.Infrastructure/Serialization/JsonStringInternConverter.cs (hunk 1 above)
Comment (Nuklon) [reply src/Umbraco.Infrastructure/Serialization/JsonStringInternConverter.cs:15]: Looking at the code of string.Intern, it still has the same downsides in .NET 8.0. I think the recommendation there to use a custom collection would be better.

File: src/Umbraco.Infrastructure/Serialization/JsonFuzzyBooleanConverter.cs (hunk 2 above)
Comment (Nuklon) [reply src/Umbraco.Infrastructure/Serialization/JsonFuzzyBooleanConverter.cs:9]: ConvertToXConverter feels a bit repetitive. Why not just JsonBooleanConverter?

File: src/Umbraco.Infrastructure/Serialization/JsonDictionaryStringInternIgnoreCaseConverter.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+using System.Text.Json;
+
+namespace Umbraco.Cms.Infrastructure.Serialization;
+
+/// <summary>
+/// Converts a dictionary with a string key to or from JSON, using the <see cref="StringComparer.OrdinalIgnoreCase" /> comparer and interning the string key when reading.
+/// </summary>
+/// <typeparam name="TValue">The type of the dictionary value.</typeparam>
+public sealed class JsonDictionaryStringInternIgnoreCaseConverter<TValue> : ReadOnlyJsonConverter<Dictionary<string, TValue>>
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Serialization/JsonDictionaryStringInternIgnoreCaseConverter.cs:9]: Same comments here about string.Internal.

File: src/Umbraco.Infrastructure/Serialization/JsonDictionaryStringIgnoreCaseConverter.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                return dictionary;
+            }
+
+            // Get key
+            if (reader.TokenType != JsonTokenType.PropertyName)
+            {
+                throw new JsonException();
+            }
+
+            string propertyName = reader.GetString() ?? throw new JsonException();
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Serialization/JsonDictionaryStringIgnoreCaseConverter.cs:34]:
```suggestion
            if (reader.TokenType != JsonTokenType.PropertyName || reader.GetString() is not { } propertyName)
            {
                throw new JsonException();
            }
```

File: src/Umbraco.PublishedCache.NuCache/DataSource/MessagePackDictionaryStringInternIgnoreCaseFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace Umbraco.Cms.Infrastructure.PublishedCache.DataSource;
+
+/// <summary>
+/// A MessagePack formatter (deserializer) for a string key dictionary that uses <see cref="StringComparer.OrdinalIgnoreCase" /> for the key string comparison and interns the string.
+/// </summary>
+/// <typeparam name="TValue">The type of the value.</typeparam>
+public sealed class MessagePackDictionaryStringInternIgnoreCaseFormatter<TValue> : DictionaryFormatterBase<string, TValue, Dictionary<string, TValue>, Dictionary<string, TValue>.Enumerator, Dictionary<string, TValue>>
+{
+    /// <inheritdoc />
+    protected override void Add(Dictionary<string, TValue> collection, int index, string key, TValue value, MessagePackSerializerOptions options)
```
Comment (Nuklon) [inline src/Umbraco.PublishedCache.NuCache/DataSource/MessagePackDictionaryStringInternIgnoreCaseFormatter.cs:13]: Same comments regarding string.Intern.

**Comment by bergmania** [comment]
Most parts make sense to me, but personally I do not understand why 
`DefaultJsonSerializer` should be better than `SystemTextJsonSerializer`(Same with `DefaultConfigurationEditorJsonSerializer` and `SystemTextConfigurationEditorJsonSerializer`).

I would prefer those to keep the names.

Regarding interning or not, I think that part should be kept in a separate PR. I'm sure more people have opinions about that.

**Comment by ronaldbarendse** [comment]
> Most parts make sense to me, but personally I do not understand why `DefaultJsonSerializer` should be better than `SystemTextJsonSerializer`(Same with `DefaultConfigurationEditorJsonSerializer` and `SystemTextConfigurationEditorJsonSerializer`).

We don't ship different implementations, so prepending them with `Default` clearly communicates that. These also aren't `SystemText` implementations of the `IJsonSerializer` and `IConfigurationEditorJsonSerializer`, but naming them `SystemTextJsonJsonSerializer` and `SystemTextJsonConfigurationEditorJsonSerializer` seemed a bit too verbose.

> Regarding interning or not, I think that part should be kept in a separate PR. I'm sure more people have opinions about that.

I agree! String interning can still improve performance and lower memory usage, but maybe using a custom collection as @Nuklon suggested would be a better alternative...

**Comment by bergmania** [comment]
@ronaldbarendse I have reverted the renaming of these classes. SystemText* is a good name indicating what technology is named. The verbose class name is not really exposed, because people use the technology agnostic interface name, when they inject.

For future PRs, I will recommend to keep to the changes you need and that blocks you, so silly things like naming are not blocking the PR :)

**Review by bergmania — APPROVED** [review]

**Pull request merged**